xdsl/dialects/hw.py

```diff
@@ -24,6 +26,34 @@
 from xdsl.printer import Printer
 
 
+@dataclass
+class InnerSymTarget:
+    """The target of an inner symbol, the entity the symbol is a handle for."""
+
+    op: Operation | None = None
+    field_id: int = 0
+    port_idx: int | None = None
+
+    def __bool__(self):
+        # None-valued op defines an invalid target
+        return self.op is not None
+
+    def is_port(self) -> bool:
+        return self.port_idx is not None
+
+    def is_field(self) -> bool:
+        return self.field_id != 0
+
+    def is_op_only(self) -> bool:
+        return not self.is_field() and not self.is_port()
+
+    @classmethod
+    def get_target_for_subfield(
```

Comment:
nit: would it make sense to add this (or appropriately modified) comment from CIRCT here as docstring?

```py
"""
Return a target to the specified field within the given base.
`field_id` is relative to the specified base target.
"""
```